### PR TITLE
[BEAM-5464] Use BATCH_FORCED as the default ExecutionMode for batch pipeline

### PR DESCRIPTION
**Pull request by angoenka**
Use BATCH_FORCED ExecutionMode for Flink batch pipelines to avoid flink scheduling dead lock.
Beam merges the chained ProcessBundles in ExecutionStage so this should not have a lot of over head.
Also introducing a parameter to change batch execution mode if needed.

Reference upstream jira https://issues.apache.org/jira/browse/FLINK-10672

**Please** add a meaningful description for your change here

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by angoenka** [comment]
cc: @mxm @robertwb

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 1]
```diff
@@ -187,4 +188,12 @@
   Long getLatencyTrackingInterval();
 
   void setLatencyTrackingInterval(Long interval);
+
+  @Description(
+      "Flink mode for data exchange for batch pipeline. "
+          + "Reference {@link org.apache.flink.api.common.ExecutionMode}")
+  @Default.Enum("BATCH_FORCED")
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: I'm a bit hesitant to change the default value here. This is also used by the non-portable FlinkRunner and the default is `PIPELINED`. We haven't heard from anyone having issues with the batch execution. I'd leave this at the Flink default until we have found out the exact issue.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 2]
```diff
@@ -187,4 +188,12 @@
...
+  ExecutionMode getExecutionModeForBatch();
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Can we add a test in `PipelineOptionsTest`?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java (hunk 1 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Also ok for me to add a link to a JIRA issue to further investigate this issue.

Comment (angoenka) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Sounds good, I will update the default.
Flink Jira for reference https://issues.apache.org/jira/browse/FLINK-10672

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java (hunk 2 above)
Comment (angoenka) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Sure

Comment (angoenka) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: TheDefault Enum test is done in ProxyInvocationHandlerTest.java
Please let me know if you are referring to some other test.

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: I meant `PipelineOptionsTest` to test the default values. I'm adding it in the merge commit.

Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Should this be getFlinkExecutionModeForBatch, as it seems rather Flink-specific?

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Considering the global namespace of the option names that would be preferred. However, not a single other option so far has a Flink prefix. Perhaps we have to come up with a way to scope option names?

**Comment by mxm** [comment]
> Just to clarify, the resulting PR just enables the option, rather than setting it as the title states, right?

Yes, it just enables to set it. The default remains unchanged (`PIPELINED`). We didn't want to make changes to the legacy FlinkRunner and further investigate the issue. As a next step we could set `BATCH_FORCED` for the portable Runner.

**Pull request merged**